File: src/File/FlacFile.cpp
/*
 * libMendel-Audio
 *
 * The MIT License (MIT)
 *
 * Copyright (C) 2017 Oskar Mendel
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "FlacFile.h"
#include <iostream>
#include <fstream>

FlacFile::FlacFile()
{

}

FlacFile::FlacFile(std::string path)
{
	filePath = path;
	//populate file with the file at the specified path.
}

FlacFile::FlacFile(FILE * f)
{
	file = f;
	//Get the file path of file and populate filePath with the path.
}

FlacFile::~FlacFile()
{
}

void FlacFile::setFile(FILE * f)
{
	file = f;
}

void FlacFile::setFilePath(std::string path)
{
	filePath = path;
}

void FlacFile::setArtist(std::string a)
{
	artist = a;
}

void FlacFile::setAlbum(std::string a)
{
	album = a;
}

void FlacFile::setTitle(std::string t)
{
	title = t;
}

void FlacFile::setGenre(std::string g)
{
	genre = g;
}

void FlacFile::setTrackNumber(std::string t)
{
	trackNumber = t;
}

void FlacFile::setDate(std::string d)
{
	date = d;
}

int FlacFile::getMinimumBlockSize()
{
	return minimumBlockSize;
}

void FlacFile::setMinimumBlockSize(int size)
{
	minimumBlockSize = size;
}

int FlacFile::getMaximumBlockSize()
{
	return maximumBlockSize;
}

void FlacFile::setMaximumBlockSize(int size)
{
	maximumBlockSize = size;
}

int FlacFile::getMinimumFrameSize()
{
	return minimumFrameSize;
}

void FlacFile::setMinimumFrameSize(int size)
{
	minimumFrameSize = size;
}

int FlacFile::getMaximumFrameSize()
{
	return maximumFrameSize;
}

void FlacFile::setMaximumFrameSize(int size)
{
	maximumFrameSize = size;
}

int FlacFile::getSampleRate()
{
	return sampleRate;
}

void FlacFile::setSampleRate(int rate)
{
	sampleRate = rate;
}

int FlacFile::getNumChannels()
{
	return numChannels;
}

void FlacFile::setNumChannels(int channels)
{
	numChannels = channels;
}

int FlacFile::getBitsPerSample()
{
	return bitsPerSample;
}

void FlacFile::setBitsPerSample(int bits)
{
	bitsPerSample = bits;
}

int FlacFile::getNumSamples()
{
	return numSamples;
}

void FlacFile::setNumSamples(int samples)
{
	numSamples = samples;
}

std::string FlacFile::getVendor()
{
	return std::string();
}

void FlacFile::setVendor(std::string v)
{
	vendor = v;
}

void FlacFile::parse()
{
	//Temp code, Reads first 4 bytes of a file, if its a flac file it should print out
	//fLaC from the output stream.
	std::ifstream in(getFilePath());

	for (int i = 0; i < 4; i++) {
		char c = in.get();
		if (in.fail()) {
			std::cout << std::endl << "Error, closing.." << std::endl;
			break;
		}
		else {
			std::cout << c;
		}
	}
	std::cout << std::endl;
	in.close();
}

File: src/main/java/me/oskarmendel/libMendelAudio/flac/FlacFile.java
/**
 * The MIT License (MIT)
 *
 * Copyright (C) 2017 Oskar Mendel
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package me.oskarmendel.libMendelAudio.flac;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;

import me.oskarmendel.libMendelAudio.MusicFile;

/**
 * Object to represent files of the Flac format.
 * Implemented using the formatting for Flac files represented here:
 * https://xiph.org/flac/format.html
 *
 * @author Oskar Mendel
 * @version 0.00.00
 * @name FlacFile.groovy
 */
public class FlacFile extends MusicFile{
	
	private int minimumBlockSize;	//Represented as 16 bits
	private int maximumBlockSize;	//Represented as 16 bits
	private int minimumFrameSize;	//Represented as 24 bits
	private int maximumFrameSize;	//Represented as 24 bits
	
	private int sampleRate;			//Represented as 20 bits
	private int numChannels;		//Represented as 3 bits
	private int bitsPerSample;		//Represented as 5 bits
	private int numSamples;			//Represented as 36 bits
	
	//Data from the Vorbis Comments within the Flac file.
	String vendor;
	
	public FlacFile() {
		
	}
	
	/**
	 * Constructs a new FlacFile object and directly populates it with data
	 * from the specified file at the target file path.
	 * 
	 * @param path - Path of the Flac file.
	 * @throws IOException 
	 * @throws FileNotFoundException 
	 */
	public FlacFile(String path) throws FileNotFoundException, IOException {
		this.filePath = path;
		this.file = new File(this.filePath);
		
		parse();
	}
	
	/**
	 * Constructs a new FlacFile object and directly populates it with data
	 * from the specified file object.
	 * 
	 * @param file - File object to retrieve data from.
	 * @throws IOException 
	 * @throws FileNotFoundException 
	 */
	public FlacFile(File file) throws FileNotFoundException, IOException {
		this.file = file;
		this.filePath = "$file";
		
		parse();
	}

	/**
	 * Initializes the parsing of data using the FlacParser class.
	 * @throws IOException 
	 * @throws FileNotFoundException 
	 * @throws IllegalArgumentException - When the FlacFile doesn't have an initialized file and filepath.
	 */
	@Override
	public void parse() throws FileNotFoundException, IOException {
		if(this.file == null || this.filePath == null) {
			throw new IllegalArgumentException("File and file path is not initialized!");
		}
		FlacParser.parseFlacFile(this.file, this);
	}

	/**
	 * @return the minimumBlockSize
	 */
	public int getMinimumBlockSize() {
		return minimumBlockSize;
	}

	/**
	 * @param minimumBlockSize the minimumBlockSize to set
	 */
	public void setMinimumBlockSize(int minimumBlockSize) {
		this.minimumBlockSize = minimumBlockSize;
	}

	/**
	 * @return the maximumBlockSize
	 */
	public int getMaximumBlockSize() {
		return maximumBlockSize;
	}

	/**
	 * @param maximumBlockSize the maximumBlockSize to set
	 */
	public void setMaximumBlockSize(int maximumBlockSize) {
		this.maximumBlockSize = maximumBlockSize;
	}

	/**
	 * @return the minimumFrameSize
	 */
	public int getMinimumFrameSize() {
		return minimumFrameSize;
	}

	/**
	 * @param minimumFrameSize the minimumFrameSize to set
	 */
	public void setMinimumFrameSize(int minimumFrameSize) {
		this.minimumFrameSize = minimumFrameSize;
	}

	/**
	 * @return the maximumFrameSize
	 */
	public int getMaximumFrameSize() {
		return maximumFrameSize;
	}

	/**
	 * @param maximumFrameSize the maximumFrameSize to set
	 */
	public void setMaximumFrameSize(int maximumFrameSize) {
		this.maximumFrameSize = maximumFrameSize;
	}

	/**
	 * @return the sampleRate
	 */
	public int getSampleRate() {
		return sampleRate;
	}

	/**
	 * @param sampleRate the sampleRate to set
	 */
	public void setSampleRate(int sampleRate) {
		this.sampleRate = sampleRate;
	}

	/**
	 * @return the numChannels
	 */
	public int getNumChannels() {
		return numChannels;
	}

	/**
	 * @param numChannels the numChannels to set
	 */
	public void setNumChannels(int numChannels) {
		this.numChannels = numChannels;
	}

	/**
	 * @return the bitsPerSample
	 */
	public int getBitsPerSample() {
		return bitsPerSample;
	}

	/**
	 * @param bitsPerSample the bitsPerSample to set
	 */
	public void setBitsPerSample(int bitsPerSample) {
		this.bitsPerSample = bitsPerSample;
	}

	/**
	 * @return the numSamples
	 */
	public int getNumSamples() {
		return numSamples;
	}

	/**
	 * @param numSamples the numSamples to set
	 */
	public void setNumSamples(int numSamples) {
		this.numSamples = numSamples;
	}

	/**
	 * @return the vendor
	 */
	public String getVendor() {
		return vendor;
	}

	/**
	 * @param vendor the vendor to set
	 */
	public void setVendor(String vendor) {
		this.vendor = vendor;
	}
}